RULES/hdl/xsim.go

```diff
@@ -52,6 +52,132 @@ var XelabDebug = core.StringFlag{
 // multiple times.
 var xsim_rules = make(map[string]bool)
 
+// Parameters of the do-file
+type tclFileParams struct {
+	DumpVcd      bool
+	DumpVcdFile  string
+}
+
+// Do-file template
+const xsim_tcl_file_template = `
+if [info exists ::env(from)] {
+	run $::env(from)
+}
+
+{{ if .DumpVcd }}
+open_vcd {{ .DumpVcdFile }}
+log_vcd [get_objects -r]
+{{ end }}
+
+if [info exists ::env(duration)] {
+	run $::env(duration)
+} else {
+	run -all
+}
+
+{{ if .DumpVcd }}
+close_vcd
+{{ end }}
+`
+
+type prjFile struct {
+	Rule   Simulation
+	Macros []string
+	Incs   []string
+	Deps   []core.Path
+	Data   []string
+}
+
+func addToPrjFile(ctx core.Context, prj prjFile, ips []Ip, srcs []core.Path) prjFile {
+	for _, ip := range ips {
+		prj = addToPrjFile(ctx, prj, ip.Ips(), ip.Sources())
+	}
+
+	for _, src := range srcs {
+		if IsHeader(src.String()) {
+			new_path := path.Dir(src.Absolute())
+			gotit := false
+			for _, old_path := range prj.Incs {
+				if new_path == old_path {
+					gotit = true
+					break
+				}
+			}
+			if !gotit {
+				prj.Incs = append(prj.Incs, new_path)
+			}
```

Comment:
I disagree. The methodology that you're suggesting (and that we're currently using) replaces a simple problem (checking for duplicate files) with a much more complicated one (tight coupling between modules).

Whenever you design a complex system you have to make sure that there is as little coupling as possible between the individual parts of the system. However, having the full path in the include statement completely breaks that. Suddenly, if, for whatever reason, we decide to move a header file in a module, we have to update the source code of every single library that uses it.

Currently, it is not a big problem, which is why I have not pushed strongly to fix it, and hopefully it will also not grow into one when we start releasing IPs to production. For now I prefer to leave the option open to use either local or global include files until we see what works better in a production environment.

Reply:
I see. Thanks for the explanation, I believe you are right.